### PR TITLE
Expunge stale entries in InternalLoggerRegistry

### LINKED ISSUE #3430: Expunge stale entries in `InternalLoggerRegistry`
`InternalLoggerRegistry` is essentially a `Map<MessageFactory, Map<String, WeakReference<Logger>>>`. [As reported](/apache/logging-log4j2/issues/3399#issuecomment-2626946587) by @tristantarrant, stale entries are not expunged in its current form. That is, the inner map does not get compacted for entries whose associated `WeakReference`s are reclaimed by the garbage collector. This might cause high memory usage if an excessive amount of ephemeral loggers get registered.

Reclaimed references can be tracked by a `ReferenceQueue` passed to the `WeakReference::new` and processed as follows:

```java
private final ReferenceQueue<Logger> staleLoggerRefs = new ReferenceQueue<>();

private void expungeStaleEntries() {
    Reference<? extends Logger> loggerRef;
    while ((loggerRef = staleLoggerRefs.poll()) != null) {
        Logger logger = loggerRef.get();
        if (logger != null) {
            removeLogger(logger);
        }
    }
}
```

When to run `expungeStaleEntries()` should be decided with care, since `removeLogger()` mutates the registry map.

### On method invocation

All `ILR` methods (`getLogger()`, `computeIfAbsent()`, etc.) can first try to expunge stale entries.
This approach is exercised by `WeakHashMap` itself and in #3427.

**Pros:**

- Implies no changes besides `ILR`

**Cons:**

- Read-only operations, which are accessed most, might try to acquire the write-lock.
- If `ILR` is not accessed, stale entries will still be kept. Consider an application where all `Logger`s are assigned, and a temporary operation creates excessive amount of loggers and completes. All created temporary `Logger` instances will be reclaimed by GC, yet will continue taking space for the associated `Map.Entry<String, WeakReference<Logger>>`.

### By a janitor thread

`ILR` can start a janitor thread listening on the `ReferenceQueue<Logger>`.

**Cons:**

- `ILR` must have a life cycle (to start/stop the janitor thread) managed by `LoggerContext`.

### PR DESCRIPTION
**Pull request by jhl221123**
This pull request addresses issue #3430 by implementing a mechanism to expunge stale entries in `InternalLoggerRegistry`. 

It builds upon PR #3474 and incorporates feedback from @vy and @ppkarwasz.

Key changes include:
*   **Enabled stale entry detection via `ReferenceQueue`**: When a `Logger` is GC'd, its `WeakReference` is enqueued. Subsequent calls to methods like `getLogger()` or `computeIfAbsent()` trigger `expungeStaleEntries()` to process it.
*   **Implemented `expungeStaleEntries` logic**: If the `ReferenceQueue` is not empty, a write lock is acquired, the queue is drained, and map entries are scanned once to remove `WeakReferences` with null referents and clean up any empty maps. No action or locking occurs if the queue is initially empty.
*   **Added unit tests**: Verified the expunging mechanism under various conditions, including GC scenarios and concurrent access.

@vy, `ReferenceQueue` has no `peek()`/`isEmpty()`, so I couldn’t do a proper double-check and used an initial `poll()` instead. Feedback is welcome if there's a better way or if I missed something. Thanks!

Fixes #3430

### CODE REVIEW

File: log4j-core-test/src/main/java/org/apache/logging/log4j/core/test/util/InternalLoggerRegistryGCTest.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+    @Test
+    void testGetLoggerReturnsNullForNonExistentLogger() {
+        assertNull(registry.getLogger("nonExistent", messageFactory));
+    }
+
+    @Test
+    void testComputeIfAbsentCreatesLogger() {
+        Logger logger = registry.computeIfAbsent(
+                "testLogger", messageFactory, (name, factory) -> loggerContext.getLogger(name, factory));
```
Comment (ppkarwasz) [inline log4j-core-test/src/main/java/org/apache/logging/log4j/core/test/util/InternalLoggerRegistryGCTest.java:74]: `loggerContext.getLogger` calls back into `InternalLoggerRegistry`, which means we're triggering nested `computeIfAbsent` calls here. Is this behavior intentional? If not, it might be safer to use a mocked `Logger` to avoid unintended side effects or recursion.

File: log4j-core-test/src/main/java/org/apache/logging/log4j/core/test/util/InternalLoggerRegistryGCTest.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            assertTrue(
+                    isExpungeStaleEntries,
+                    "Stale WeakReference entries were not removed from the inner map for MessageFactory");
+        });
+    }
+
+    @Test
+    void testExpungeStaleMessageFactoryEntry() {
+        Logger logger = registry.computeIfAbsent(
+                "testLogger", messageFactory, (name, factory) -> loggerContext.getLogger(name, factory));
```
Comment (ppkarwasz) [inline log4j-core-test/src/main/java/org/apache/logging/log4j/core/test/util/InternalLoggerRegistryGCTest.java:131]: The `messageFactory` parameter is defined as `SimpleMessageFactory.INSTANCE` and therefore can not be garbage collected. Should we use a different `messageFactory` for this test case?

**Review by ppkarwasz — CHANGES_REQUESTED** [review]
@jhl221123, thank you for picking up this pull request!

The main code looks good to me overall. I do have a few suggestions regarding the unit tests:

File: log4j-core-test/src/main/java/org/apache/logging/log4j/core/test/util/InternalLoggerRegistryGCTest.java (hunk 1 above)
Comment (jhl221123) [reply log4j-core-test/src/main/java/org/apache/logging/log4j/core/test/util/InternalLoggerRegistryGCTest.java]: Not intended! The logger supplier lambdas now use `new MockLogger(...)` directly.

File: log4j-core-test/src/main/java/org/apache/logging/log4j/core/test/util/InternalLoggerRegistryGCTest.java (hunk 2 above)
Comment (jhl221123) [reply log4j-core-test/src/main/java/org/apache/logging/log4j/core/test/util/InternalLoggerRegistryGCTest.java]: Good point! That test case now uses a locally created `new SimpleMessageFactory()` instance.

**Review by ppkarwasz — APPROVED** [review]
The changes look solid—nice work! There's just one final task before we can wrap this up: could you please add a changelog entry under `src/changelog/.2.x.x`, as outlined in the [changelog documentation](https://logging.apache.org/log4j/tools/log4j-changelog.html#changelog-entry-file)?

Also, just a quick note on `loggerRefByNameByMessageFactory`: since it's a `WeakHashMap`, it doesn’t follow all the typical `Map` invariants. From what I can tell, you're using it safely, and I didn’t spot any issues that would lead to unexpected behavior. That said, given the sensitivity of this part of the codebase, I’d like a second pair of eyes on it.

@vy, could you take a moment to review this as well?

File: log4j-core/src/main/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistry.java [hunk 3]
```diff
@@ -53,13 +58,43 @@ public final class InternalLoggerRegistry {
...
+    /**
+     * Expunges stale entries for logger references and message factories.
+     */
+    private void expungeStaleEntries() {
+        Reference<? extends Logger> loggerRef = staleLoggerRefs.poll();
+
+        if (loggerRef != null) {
+            writeLock.lock();
+            try {
+                while (staleLoggerRefs.poll() != null) {
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistry.java:78]: False positive, the comment is already there.

**Comment by vy** [comment]
I would like to and I will. Please allow me some time.

Op vr 23 mei 2025 om 21:27 schreef Piotr P. Karwasz <
***@***.***>

File: log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            throws NoSuchFieldException, IllegalAccessException {
+        Field loggerMapField = registry.getClass().getDeclaredField("loggerRefByNameByMessageFactory");
+        loggerMapField.setAccessible(true);
+        @SuppressWarnings("unchecked")
+        Map<MessageFactory, Map<String, WeakReference<Logger>>> loggerMap =
+                (Map<MessageFactory, Map<String, WeakReference<Logger>>>) loggerMapField.get(registry);
+        return loggerMap;
+    }
+
+    private class MockLogger extends Logger {
```
Comment (vy) [inline log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java:183]: I don't think you really need this definition. All your `new MockLogger(...)` calls can be replaced with `new Logger(...) {}` instead.

File: log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java
Comment (vy) [inline log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java:1]: Can you use `final` modifier for variables wherever applicable, please?

File: log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            Logger logger = registry.computeIfAbsent(
+                    loggerNamePrefix + i,
+                    messageFactory,
+                    (name, factory) -> new MockLogger(loggerContext, name, factory));
+            logger.info("Using logger {}", logger.getName());
+        }
+
+        await().atMost(10, SECONDS).pollInterval(100, MILLISECONDS).untilAsserted(() -> {
+            System.gc();
+            System.runFinalization();
```
Comment (vy) [inline log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java:108]: Don't use this deprecated-for-removal method, please.
```suggestion
```

File: log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+            boolean isExpungeStaleEntries = true;
+            for (int i = 0; i < numberOfLoggers; i++) {
+                if (loggerRefByName.containsKey(loggerNamePrefix + i)) {
+                    isExpungeStaleEntries = false;
+                    break;
+                }
+            }
+            assertTrue(
+                    isExpungeStaleEntries,
+                    "Stale WeakReference entries were not removed from the inner map for MessageFactory");
```
Comment (vy) [inline log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java:125]: Instead of this, could you _count_ the number of unexpected logger entries and assert that they are equal to zero, and if not, report how many are found, please?

File: log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    void testExpungeStaleMessageFactoryEntry() {
+        SimpleMessageFactory mockMessageFactory = new SimpleMessageFactory();
+        Logger logger = registry.computeIfAbsent(
+                "testLogger", mockMessageFactory, (name, factory) -> new MockLogger(loggerContext, name, factory));
+        logger.info("Using logger {}", logger.getName());
+        logger = null;
+
+        await().atMost(10, SECONDS).pollInterval(100, MILLISECONDS).untilAsserted(() -> {
+            System.gc();
+            System.runFinalization();
```
Comment (vy) [inline log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java:139]:
```suggestion
```

File: log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            Map<MessageFactory, Map<String, WeakReference<Logger>>> loggerRefByNameByMessageFactory =
+                    reflectAndGetLoggerMapFromRegistry();
+            assertNull(
+                    loggerRefByNameByMessageFactory.get(mockMessageFactory),
+                    "Stale MessageFactory entry was not removed from the outer map");
+        });
+    }
+
+    @Test
+    void testConcurrentAccess() throws InterruptedException {
```
Comment (vy) [inline log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java:151]: @jhl221123, @ppkarwasz, do we have a inspiration source (scenario, bug report, etc.) for this particular test?

**Review by vy — CHANGES_REQUESTED** [review]

File: log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java (hunk 8 above)
Comment (jhl221123) [reply log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java]: @vy, Thanks for the review!
I couldn't think of a specific scenario, so I've removed it for now. Please let me know if anything changes!

Comment (ppkarwasz) [reply log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/internal/InternalLoggerRegistryTest.java]: We've previously seen issues with `InternalLoggerRegistry` under concurrent usage—see #3252 and #3399—caused by recursive calls from third-party code. I believe this was addressed in #3418, though I didn’t add a regression test due to the difficulty of reproducing static initialization edge cases.

That said, since `InternalLoggerRegistry` replaced `ConcurrentHashMap` (multi-lock) with a single lock, it might be worth adding a benchmark to assess any performance regression.

**Review by vy — APPROVED** [review]
